app/scopes/preload_helper.ts

```diff
@@ -0,0 +1,79 @@
+import { scope } from "@adonisjs/lucid/orm";
+import {
+  LucidModel,
+  LucidRow,
+  ModelQueryBuilderContract,
+} from "@adonisjs/lucid/types/model";
+import { ExtractModelRelations } from "@adonisjs/lucid/types/relations";
+
+/**
+ * Preloads specified relations on a Lucid model query.
+ *
+ * The returned scope function accepts an object where:
+ * - **Keys** are relation names, which can be nested using dot notation (e.g., `'campus.building'`).
+ * - **Values** are booleans indicating whether to preload the corresponding relation.
+ *
+ * @template T - Type of the Lucid model.
+ * @param {T} model - The Lucid model to preload relations on.
+ * @returns {QueryScope} A scope function that takes a query and a relations object.
+ *
+ * @example
+ * const buildings = await Building.query()
+ *  .withScopes((scopes) => {
+ *    scopes.preloadRelations(request.only(["campus"]));
+ */
+export const preloadRelations = <T extends LucidModel>(model: T) =>
+  scope(
+    (
+      query: ModelQueryBuilderContract<LucidModel, LucidRow>,
+      relations: Partial<Record<string, string>>,
+    ) => {
+      for (const relation in relations) {
+        const value = relations[relation];
+
+        if (value === "1" || value === "true") {
+          query = preloadSinglePath(query, model, relation.split("."));
+        }
+      }
+      return query;
+    },
+  );
+
+/**
+ * Helper function to work with recursively looaded chained relations
+ * - Base case [1] of function is the exhaustion of relationParts array (stack)
+ * - Base case [2] of function is the mismatch in provided relation name
+ *
+ * @param query - current query to work on
+ * @param {T} model - The Lucid model to validate NEXT related model preload
+ * @param relationParts - array that holds further chained relations that remain to be loaded in that single path
+ * @returns {ModelQueryBuilderContract<T>} query to chain other queries on
+ */
+function preloadSinglePath<T extends LucidModel>(
+  query: ModelQueryBuilderContract<T>,
+  model: T,
+  relationParts: string[],
+): ModelQueryBuilderContract<T> {
+  const relation = relationParts.shift();
+  if (relation === undefined) {
+    return query;
+  }
+
+  const relationDefinition = model.$relationsDefinitions.get(relation);
+  if (relationDefinition === undefined) {
+    return query;
+  }
```

Comment:
> Is this `relationDefinition === undefined` a case where the string that someone put in was invalid/had an invalid relation? 

Yes, my first idea was to simply ignore it bc I thought any error status response is not needed. The lack of preloaded data
sent to user is self explanatory. Do you mean any particular way of logging it ?

Reply:
I see that the controllers filter out only allowed relations from the request to pass down to the scope, so it's possible that someone might make a typo while creating those relation allowlists.
I was thinking that something as simple as a `console.log` (or another logging equivalent) with a message that an invalid relation was given would work.
This would mainly be useful when testing/debugging your own controllers locally.